src/consumer.rs

```diff
@@ -0,0 +1,306 @@
+    client: SharedFluvio,
+    topic: String,
+    partition: i32,
+}
+
+impl PartitionConsumerWrapper {
+    pub fn new(client: SharedFluvio, topic: String, partition: i32) -> Self {
+        Self {
+            client,
+            topic,
+            partition,
+        }
+    }
+}
+
+impl TryIntoJs for PartitionConsumerWrapper {
+    fn try_to_js(self, js_env: &JsEnv) -> Result<napi_value, NjError> {
+        debug!("converting FluvioWrapper to js");
+        let new_instance = PartitionConsumerJS::new_instance(js_env, vec![])?;
+        debug!("instance created");
+        PartitionConsumerJS::unwrap_mut(js_env, new_instance)?.set_client(self.client);
+        PartitionConsumerJS::unwrap_mut(js_env, new_instance)?.set_topic(self.topic);
+        PartitionConsumerJS::unwrap_mut(js_env, new_instance)?.set_partition(self.partition);
+        Ok(new_instance)
+    }
+}
+
+pub struct PartitionConsumerJS {
+    inner: Option<SharedFluvio>,
+    topic: Option<String>,
+    partition: Option<i32>,
+}
+
+#[node_bindgen]
+impl PartitionConsumerJS {
+    #[node_bindgen(constructor)]
+    pub fn new() -> Self {
+        Self {
+            inner: None,
+            topic: None,
+            partition: None,
+        }
+    }
+
+    pub fn set_client(&mut self, client: SharedFluvio) {
+        self.inner.replace(client);
+    }
+
+    pub fn set_topic(&mut self, topic: String) {
+        self.topic.replace(topic);
+    }
+
+    pub fn set_partition(&mut self, partition: i32) {
+        self.partition.replace(partition);
+    }
+
+    #[node_bindgen]
+    async fn fetch(
+        &self,
+        offset: OffsetWrapper,
+    ) -> Result<FetchablePartitionResponseWrapper, FluvioError> {
+        let topic = self.topic.clone().unwrap_or_default();
+        let partition = self.partition.unwrap_or_default();
+
+        if let Some(client) = self.inner.clone() {
+            let client: PartitionConsumer = client
+                .write()
+                .await
+                .partition_consumer(topic, partition)
+                .await?;
+
+            let response = client.fetch(offset.0).await?;
+            Ok(FetchablePartitionResponseWrapper(response))
+        } else {
+            Err(FluvioError::Other("failed to lock client".to_owned()))
+        }
+    }
+
+    #[node_bindgen(mt)]
+    async fn stream<F: Fn(String, RecordSetWrapper) + 'static + Send + Sync>(
+        &self,
+        offset: OffsetWrapper,
+        cb: F,
+    ) -> Result<(), FluvioError> {
+        let topic = self.topic.clone().unwrap_or_default();
+        let partition = self.partition.unwrap_or_default();
+
+        if let Some(client) = self.inner.clone() {
+            let client: PartitionConsumer = client
+                .write()
+                .await
+                .partition_consumer(topic, partition)
+                .await?;
+
+            let handle = spawn(PartitionConsumerJS::stream_inner(client, offset, cb)).await;
+
+            if let Err(e) = handle {
```

Comment:
What is the control-flow semantics here? I thought the point of `spawn` was to release control-flow to a separate task and continue along, but here it seems like we need to wait on the result anyway?

Reply:
This was left over from debugging an error due to an async callback. We don't need to wait for a result, this is only being handled if there is an error in the spawned thread. We can keep it for debugging or we can remove it and not await for an error.

Reply:
let's remove it

Reply:
Commit db85cf3 removes.